url2114d537a2_ForNeVeR_lanterna_tgz-pJ8-com_googlecode_lanterna_test_gui_DifferentKindsOfListBoxesTestJ8/src/com/googlecode/lanterna/terminal/InputEnabledAbstractTerminal.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * This file is part of lanterna (http://code.google.com/p/lanterna/).
+ * 
+ * lanterna is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * 
+ * Copyright (C) 2010-2012 Martin
+ */
+
+package com.googlecode.lanterna.terminal;
+
+import com.googlecode.lanterna.LanternaException;
+import com.googlecode.lanterna.input.InputDecoder;
+import com.googlecode.lanterna.input.InputProvider;
+import com.googlecode.lanterna.input.Key;
+import com.googlecode.lanterna.input.KeyMappingProfile;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.Queue;
+
+/**
+ * This is an abstract terminal that can also read input events (keys), with a
+ * default implementation of the methods from {@code InputProvider}.
+ * @author Martin
+ */
+public abstract class InputEnabledAbstractTerminal extends AbstractTerminal implements InputProvider {
+    private final InputDecoder inputDecoder;
+    private final Queue<Key> keyQueue;
+    private final Object readMutex;
+
+    public InputEnabledAbstractTerminal(InputDecoder inputDecoder) {
+        this.inputDecoder = inputDecoder;
+        this.keyQueue = new LinkedList<Key>();
+        this.readMutex = new Object();
+    }
+    
+    @Override
+    public void addInputProfile(KeyMappingProfile profile) {
+        inputDecoder.addProfile(profile);
+    }
+    
+    protected TerminalSize waitForTerminalSizeReport(int timeoutMs) {
+        long startTime = System.currentTimeMillis();
+        synchronized(readMutex) {
+            while(System.currentTimeMillis() - startTime < timeoutMs) {
+                Key key = inputDecoder.getNextCharacter();
+                if(key == null) {
+                    try {
+                        Thread.sleep(1);
+                    }
+                    catch(InterruptedException e) {}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)